elastic/beats
# Autodiscover

This PR adds support for Autodiscover, and implements a first provider for Docker.

### Autodiscover

Autodiscover allows the user to define different providers to watch for system changes and instantiate new modules as these happen.

Providers watch for changes and emit events to a common bus, then the autodiscovery process detect situations when there is something we can monitor and instantiate a modules for it.

Bus messages contain useful information about what providers see, with some common well-known optional fields (provider specific fields may also be present):

- host - service host
- port - service port
- config - recommended config for this instance (from the user or environment)

#### Docker provider

Docker provider watches for docker container events, it supports config mapping from container metadata -> config templates, so new modules are created when a container starts. The available data for a container (that you can use in the config template) is:

- host
- port
- docker.container.id
- docker.container.name
- docker.container.image
- docker.container.labels

#### Testing

This should be enough to monitor any present and future redis instances that run on docker:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            equals.docker.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.port}"
```

Or reading redis logs only:
```
filebeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            equals.docker.container.image: redis
          config:
            - type: log
              paths:
                - /var/lib/docker/containers/${data.docker.container.id}/*.log
```

You can launch it like this:
```
docker run -p 6379:6379 redis
```
This PR is in progress, with some missing parts:
- [x] Revisit Docker IP/port mapping, do we really need exposed ports? probably not
- [x] Add support in filebeat
- [x] Revisit bus implementation, use interfaces (instead of MapStr)?
- [x] Add docker metadata to modules discovered by docker provider
- [x] Update metricbeat and filebeat YAML files
